### PR TITLE
fix(net): three bugs in isLocalIP, including a crafted hostname reading as local

### PR DESCRIPTION
**Pull request by bassings**
`isLocalIP()` had **three** separate bugs, each invisible to the others. Only the first was reported by the scanner, and the third is the most serious.

Found while working the MEDIUM-and-above sweep, in the set the earlier volume-ranked pass never reached.

## Why it matters

The only caller is `couchpotato/core/http_client.py:131`:

```python
if count > MAX_FAILURES_BEFORE_DISABLE and not isLocalIP(host):
```

That exempts local hosts from being **permanently disabled** after repeated failures. Getting it wrong in one direction disables your own local services; getting it wrong in the other exempts hosts that should never be exempt.

## Bug 1: a JavaScript regex literal pasted into Python

```python
regex = r'/(^127\.)|(^192\.168\.)|...|(^::1)$/'
```

In JavaScript the wrapping `/.../` delimits the pattern. In Python they are literal characters, so two of the seven alternatives can never match: `/` cannot precede `^`, and cannot follow `$`. `::1` never matched, and `127.` matched only through a fallback.

## Bug 2: `str.lstrip` takes a character set, not a prefix

**This is why fixing bug 1 alone would not have fixed anything.**

```python
ip = ip.lstrip('htps:/')     # ':' is in the set
'::1'.lstrip('htps:/')       # -> '1'
```

The loopback marker was eaten before the regex ever ran. The same line turned `host.local` into `ost.local`.

## Bug 3: start-anchored IPv4, so a crafted hostname read as local

Found only because round two demanded an explicit list of inputs that must stay `False`:

```
isLocalIP('127.0.0.1.evil.com')   ->  True     (before)
```

`^127\.` matches a hostname that merely BEGINS with `127.`, and the redundant `ip[:4] == '127.'` fallback had the identical flaw. **This predates this PR entirely** and had nothing to do with the finding that started it. `127.0.0.1.evil.com` is a registerable domain, and this function decides which hosts skip being disabled. Both are now fully anchored and the redundant fallback is gone.

## Round one was wrong, and a green gate said nothing

The first fix passed 15 tests and a full gate, and **did not fix the bug**. Review caught it.

`http_client.py:203` builds what it passes as `f'{hostname}{":"+port if port else ""}'`, and `urlparse` strips brackets from IPv6 hosts. So `http://[::1]:9117/api` arrives as `'::1:9117'`, and the end-anchored `^::1$` never matched. A self-hosted local service almost always has an explicit port, so that is the common case rather than an edge case.

Every one of those 15 tests used a hand-written bare address. Not one built the string the way the caller does. The suite was green while the real code path stayed broken.

There is now a dedicated test class, `TestIsLocalIPMatchesHttpClientHostShape`, that constructs the host via `urlparse` exactly as the caller does. That is the gap that let it through.

## Behaviour, measured across all shapes

| input | before | after |
|---|---|---|
| `http://[::1]:9117/api` -> `::1:9117` | **False** | **True** |
| `::1` | False | True |
| `0:0:0:0:0:0:0:1` | False | True |
| `[::1]` / `[::1]:9117` | False | True |
| `127.0.0.1` / `127.0.0.1:9117` | True | True |
| `192.168.1.5:8080`, `10.0.0.5`, `172.16.0.1`, `172.31.255.1` | True | True |
| **`127.0.0.1.evil.com`** | **True** | **False** |
| `127.0.0.1:evil` | True | False |
| `172.15.0.1` / `172.32.0.1` | False | False |
| `8.8.8.8` / `8.8.8.8:443` | False | False |
| `2001:db8::1` / `2001:db8::1:9117` | False | False |
| `evil.com:127.0.0.1`, `not-127.0.0.1.example.com` | False | False |

## Deliberate decision on an ambiguous case

`::1:9117` is genuinely ambiguous as a bare string: loopback with a port, or an address in its own right. The rule chosen and documented in the code: a trailing `:<digits>` is read as a port **only if** stripping it leaves an exact recognised loopback form. Otherwise the whole string is treated as the address. That keeps `2001:db8::1:9117` remote.

General IPv6 canonicalisation (case folding, partial zero-compression, mixed forms) is deliberately out of scope. This helper recognises loopback; it does not parse arbitrary IPv6.

## Verification

Suite 3896 to 3927 passed, 2 skipped and 5 xfailed unchanged, so the delta is exactly the 31 new tests. Full local gate green including E2E. Ruff clean.

Four mutations, each confirmed landed by hash, each red for the right reason, each restored by file copy and confirmed byte-identical:

1. IPv4 `$` anchor removed, broke the `127.0.0.1.evil.com` and `127.0.0.1:evil` guards
2. bracket handling disabled, broke `[::1]` and `[::1]:9117`
3. the "recognised form" qualifier dropped from the ambiguous-port rule, broke the bare loopback forms
4. exact match changed to substring containment, broke `2001:db8::1` and `2001:db8::1:9117`

Mutation 3 is worth recording: the prediction of which test would catch it was **wrong**. It was expected to break `2001:db8::1:9117` and instead broke the bare loopback cases. The guard was load bearing either way, but via a different path than predicted, which is the sort of thing worth reporting rather than quietly rewriting.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01Ss6TA5seR8ykyJfe3itaSc

### CODE REVIEW

File: couchpotato/core/helpers/variable.py [hunk 1]
```diff
@@ -171,8 +171,31 @@ def sha256(text):
...
+    # (start of string) or immediately after '$' (end of string). The
+    # practical effect was that IPv6 loopback ('::1') was never recognised.
+    #
+    # '::1' and its uncompressed equivalent '0:0:0:0:0:0:0:1' are both
+    # matched deliberately, since a client may hand us either form of the
+    # same address. General IPv6 canonicalisation (case folding, partial
+    # zero-compression, mixed forms) is deliberately out of scope here --
+    # this helper only needs to recognise the loopback address, not parse
+    # arbitrary IPv6.
+    regex = r'(^127\.)|(^192\.168\.)|(^10\.)|(^172\.1[6-9]\.)|(^172\.2[0-9]\.)|(^172\.3[0-1]\.)|(^0:0:0:0:0:0:0:1$)|(^::1$)'
```
[the remark this answers is not in the document]
Comment (bassings) [reply couchpotato/core/helpers/variable.py]: You are right, and the fix in round one did not fix the bug. Confirmed by reproducing your exact case before changing anything:

    http://[::1]:9117/api  ->  host="::1:9117"  ->  isLocalIP = False

Every one of my 15 tests used a hand-written bare address. Not one built the string the way `http_client.py:203` does. So the suite was green while the real code path stayed broken, which is the failure mode this project keeps producing and which I had spent the afternoon describing.

Fixed in 48d1a83b8. `isLocalIP` now extracts the address from three shapes before matching: bare, `host:port`, and bracketed IPv6 (`[::1]`, `[::1]:9117`). The genuinely ambiguous case is resolved deliberately and the rule is written into the code: a trailing `:<digits>` is read as a port ONLY if stripping it leaves an exact recognised loopback form, so `2001:db8::1:9117` stays remote.

There is now a dedicated test class, `TestIsLocalIPMatchesHttpClientHostShape`, that builds the host via `urlparse` exactly as the caller does. That is the gap that let this through, so the guard now sits at the boundary rather than on the function in isolation.

**Your finding also surfaced a third, worse bug.** Writing the required must-stay-False cases showed the IPv4 alternatives were only start-anchored:

    isLocalIP("127.0.0.1.evil.com")  ->  True

A registerable domain was being classified as local, in the function that decides which hosts are exempt from being disabled after repeated failures. That predates this PR entirely and no scanner reported it. Both are now fully anchored, and the redundant `ip[:4] == "127."` fallback, which shared the same flaw, is gone.

Verified across 25 cases including all four of your URL shapes and eleven hostile inputs. Suite 3896 to 3927, skips and xfails unchanged. Four mutations, each landed-by-hash, red for the right reason, restored byte-identical.

[the remark this answers is not in the document]
Comment (bassings) [reply couchpotato/core/helpers/variable.py]: You are right, and the fix in round one did not fix the bug. Confirmed by reproducing your exact case before changing anything:

    http://[::1]:9117/api  ->  host="::1:9117"  ->  isLocalIP = False

Every one of my 15 tests used a hand-written bare address. Not one built the string the way `http_client.py:203` does. So the suite was green while the real code path stayed broken, which is the failure mode this project keeps producing and which I had spent the afternoon describing.

Fixed in 48d1a83b8. `isLocalIP` now extracts the address from three shapes before matching: bare, `host:port`, and bracketed IPv6 (`[::1]`, `[::1]:9117`). The genuinely ambiguous case is resolved deliberately and the rule is written into the code: a trailing `:<digits>` is read as a port ONLY if stripping it leaves an exact recognised loopback form, so `2001:db8::1:9117` stays remote.

There is now a dedicated test class, `TestIsLocalIPMatchesHttpClientHostShape`, that builds the host via `urlparse` exactly as the caller does. That is the gap that let this through, so the guard now sits at the boundary rather than on the function in isolation.

**Your finding also surfaced a third, worse bug.** Writing the required must-stay-False cases showed the IPv4 alternatives were only start-anchored:

    isLocalIP("127.0.0.1.evil.com")  ->  True

A registerable domain was being classified as local, in the function that decides which hosts are exempt from being disabled after repeated failures. That predates this PR entirely and no scanner reported it. Both are now fully anchored, and the redundant `ip[:4] == "127."` fallback, which shared the same flaw, is gone.

Verified across 25 cases including all four of your URL shapes and eleven hostile inputs. Suite 3896 to 3927, skips and xfails unchanged. Four mutations, each landed-by-hash, red for the right reason, restored byte-identical.
